File: leetcode/201-300/249-group-shifted-strings/bench/group_shifted.py
"""Benchmark workload for LeetCode #249 — Group Shifted Strings (Python; scale lane)."""


def canonical(word):
    b = word.encode()
    n = len(b)
    if n == 0:
        return ""
    shift = b[0] - ord('a')
    out = []
    for i in range(n):
        c = ((b[i] - ord('a') - shift) + 26) % 26
        out.append(f"{c},")
    return "".join(out)


def main():
    words_n = 120000
    rounds = 5

    words = []
    state = 249249
    for _ in range(words_n):
        state = (state * 1103515245 + 12345) & 2147483647
        ln = (state // 65536) % 10 + 3
        state = (state * 1103515245 + 12345) & 2147483647
        seed = (state // 65536) % 40
        state = (state * 1103515245 + 12345) & 2147483647
        shift = (state // 65536) % 26

        s = []
        for i in range(ln):
            base = (seed * 7 + i * 11) % 26
            ch = (base + shift) % 26
            s.append(chr(97 + ch))
        words.append("".join(s))

    sink = 0
    for _ in range(rounds):
        table = {}
        groups = 0
        keysum = 0
        for w in words:
            key = canonical(w)
            for c in key.encode():
                keysum = (keysum * 31 + c) % 1000000007
            if key not in table:
                groups += 1
            table.setdefault(key, []).append(w)
        sink = (sink * 131 + groups) % 1000000007
        sink = (sink * 31 + keysum) % 1000000007
    print(sink)


main()

File: leetcode/201-300/249-group-shifted-strings/bench/group_shifted.rs
// Benchmark workload for LeetCode #249 — Group Shifted Strings (Rust mirror).
// Mirrors group_shifted.kara algorithm-for-algorithm, including the two map
// probes per word (contains_key, then entry) that the Kara version performs.
use std::collections::HashMap;

fn canonical(word: &str) -> String {
    let bytes = word.as_bytes();
    let n = bytes.len();
    if n == 0 {
        return String::new();
    }
    let shift = bytes[0] as i64 - b'a' as i64;
    let mut out = String::new();
    for i in 0..n {
        let c = ((bytes[i] as i64 - b'a' as i64 - shift) + 26) % 26;
        out.push_str(&format!("{},", c));
    }
    out
}

fn main() {
    let words_n: i64 = 120000;
    let rounds: i64 = 5;

    let mut words: Vec<String> = Vec::new();
    let mut state: i64 = 249249;
    for _ in 0..words_n {
        state = (state * 1103515245 + 12345) & 2147483647;
        let len = (state / 65536) % 10 + 3;
        state = (state * 1103515245 + 12345) & 2147483647;
        let seed = (state / 65536) % 40;
        state = (state * 1103515245 + 12345) & 2147483647;
        let shift = (state / 65536) % 26;

        let mut s = String::new();
        for i in 0..len {
            let base = (seed * 7 + i * 11) % 26;
            let ch = (base + shift) % 26;
            s.push((97 + ch) as u8 as char);
        }
        words.push(s);
    }

    let mut sink: i64 = 0;
    for _ in 0..rounds {
        let mut table: HashMap<String, Vec<String>> = HashMap::new();
        let mut groups: i64 = 0;
        let mut keysum: i64 = 0;
        for w in &words {
            let key = canonical(w);
            for &c in key.as_bytes() {
                keysum = (keysum * 31 + c as i64) % 1000000007;
            }
            if !table.contains_key(&key) {
                groups += 1;
            }
            table.entry(key).or_insert_with(Vec::new).push(w.clone());
        }
        sink = (sink * 131 + groups) % 1000000007;
        sink = (sink * 31 + keysum) % 1000000007;
    }
    println!("{}", sink);
}
